google-http-client-apache-v5/src/main/java/com/google/api/client/http/apache/v5/Apache5ResponseContent.java

```diff
@@ -0,0 +1,36 @@
+package com.google.api.client.http.apache.v5;
+
+import java.io.IOException;
+import java.io.InputStream;
+import org.apache.hc.core5.http.ClassicHttpResponse;
+
+public class Apache5ResponseContent extends InputStream {
+  private final ClassicHttpResponse response;
+  private final InputStream wrappedStream;
+
+  public Apache5ResponseContent(InputStream wrappedStream, ClassicHttpResponse response) {
+    this.response = response;
+    this.wrappedStream = wrappedStream;
+  }
+
+  @Override
+  public int read() throws IOException {
```

Comment:
there are several overloads of `read()`, but all of them are ways to organize groups of calls to this exact function ([1](https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/19fb8f93c59dfd791f62d41f332db9e306bc1422/src/java.base/share/classes/java/io/InputStream.java#L271), [2](https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/19fb8f93c59dfd791f62d41f332db9e306bc1422/src/java.base/share/classes/java/io/InputStream.java#L280)), hence us not needing to override the other methods

Reply:
This class needs to override all the methods of `InputStream` and delegate them to the `wrappedStream`. 

We don't know what subclass (or nested subclasses) of `InputStream` we're working with and what they've modified. By not delegating to the `wrappedStream` implementation for each method, we are in effect overriding the wrappedStream's implementations and forcing them back to the default implementations in the base `InputStream` class.

For example: Since `available()` isn't overridden, we've now ensured that it [always returns 0](https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/19fb8f93c59dfd791f62d41f332db9e306bc1422/src/java.base/share/classes/java/io/InputStream.java#L568) regardless of what the wrappedStream would have returned.

Finally, we could reason about which methods exist that might be an exception to this rule (as suggested above: maybe we don't override + delegate all the read methods) but this introduces the possibility that we're wrong when the solution to ensure we can't be wrong is very simple: override + delegate to all of them.

Reply:
TIL, makes total sense, thanks!

> but this introduces the possibility that we're wrong when the solution to ensure we can't be wrong is very simple: override + delegate to all of them.

Indeed it's simpler to avoid this by having full control over the implemented methods.